### PR TITLE
from theory to practice implementing totp 2fa in python and flask

### PR DESCRIPTION
**Pull request by prince-joel**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by cyberShaw** [comment]
@prince-joel could you dm me an avatar image of yourself on slack that I can upload it?

File: articles/from-theory-to-practice-implementing-totp-2fa-in-python-and-flask/index.md [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+* It boosts customer’s confidence in an organization.
+* It ensures only the intended user can assess the protected information.
+
+### Time-based One-Time Password (TOTP)
+
+Time-based One-Time Password (TOTP) is a common way of implementing two-factor authentication in applications. It works by asking the user for a token usually sent in an SMS, email, or generated secret pass to the user’s device with an expiry time. It compares the provided token with the actual generated token, then authenticates them if the tokens match.
+
+### Google Authenticator
+
+Google Authenticator uses a software-based authentication technique made by google that implements 2FA using TOTP and HMAC-based one-time password (HOTP) for authenticating users of an application. It is part of [Open Authentication](https://oauth.net/2/) (OATH).
```
Comment (cyberShaw) [inline articles/from-theory-to-practice-implementing-totp-2fa-in-python-and-flask/index.md]: > HMAC-based one-time password

Please expand HMAC and give a brief two line definition on what it is.

File: articles/from-theory-to-practice-implementing-totp-2fa-in-python-and-flask/index.md [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+pip install pyotp
+pip install flask-bootstrap4
+```
+
+#### Building a Simple Flask Server
+
+You will write the code for setting up the Flask server. Start by creating a file named `app.py` and save the code below in it:
+
+```python
+from flask import *
```
Comment (cyberShaw) [inline articles/from-theory-to-practice-implementing-totp-2fa-in-python-and-flask/index.md]: Please include comments in the code that explain what is being done

This makes it easier for a beginner to follow

File: articles/from-theory-to-practice-implementing-totp-2fa-in-python-and-flask/index.md [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+You should get an image similar to the one below when an invalid token is provided:
+
+![invalid OTP](\engineering-education\from-theory-to-practice-implementing-totp-2fa-in-python-and-flask\invalid OTP.png)
+
+You should get an image similar to the one below when a valid token is provided:
+
+![valid OTP](\engineering-education\from-theory-to-practice-implementing-totp-2fa-in-python-and-flask\valid OTP.png)
+
+### Conclusion
+
```
Comment (cyberShaw) [inline articles/from-theory-to-practice-implementing-totp-2fa-in-python-and-flask/index.md]: Please include a GitHub link of the final code of the application that you built.

**Review by cyberShaw — CHANGES_REQUESTED** [review]

**Comment by prince-joel** [comment]
@cyberShaw I have added the avatar image

**Comment by cyberShaw** [comment]
The Avatar looks good 👍
Also, please fill out the checkboxes in the PR description

**Comment by prince-joel** [comment]
@cyberShaw I have filled out the checkboxes in the PR description

**Comment by cyberShaw** [comment]
Please make the changes that I've suggested in order to move on to the next round of review.

**Comment by prince-joel** [comment]
cybershaw

**Comment by prince-joel** [comment]
@cyberShaw  i have made the changes you requested

**Review by cyberShaw — APPROVED** [review]

**Comment by hectorkambow** [comment]
plagiarism check done

**Pull request merged**